src/Orleans.Runtime/Hosting/EndpointOptionsExtensions.cs

```diff
@@ -0,0 +1,74 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Net;
+using System.Net.Sockets;
+
+namespace Orleans.Hosting
+{
+    internal static class EndpointOptionsExtensions
+    {
+        private static IPAddress ResolveIPAddress(this EndpointOptions options)
+        {
+            if (options.IPAddress != null)
+                return options.IPAddress;
+
+            var family = options.NetworkAddress.AddressFamily;
+            var loopback = family == AddressFamily.InterNetwork ? IPAddress.Loopback : IPAddress.IPv6Loopback;
+            var any = family == AddressFamily.InterNetwork ? IPAddress.Any : IPAddress.IPv6Any;
+
+            // IF the address is an empty string, default to the local machine
+            if (string.IsNullOrEmpty(options.HostNameOrIPAddress))
+            {
+                options.HostNameOrIPAddress = Dns.GetHostName();
+            }
+
+            // Fix StreamFilteringTests_SMS tests
+            if (options.HostNameOrIPAddress.Equals("loopback", StringComparison.OrdinalIgnoreCase))
+            {
+                return loopback;
+            }
+
+            // check if addrOrHost is a valid IP address including loopback (127.0.0.0/8, ::1) and any (0.0.0.0/0, ::) addresses
+            IPAddress address;
+            if (IPAddress.TryParse(options.HostNameOrIPAddress, out address))
+            {
+                return address;
+            }
+
+            var candidates = new List<IPAddress>();
+
+            // Get IP address from DNS. If addrOrHost is localhost will 
+            // return loopback IPv4 address (or IPv4 and IPv6 addresses if OS is supported IPv6)
+            var nodeIps = Dns.GetHostAddresses(options.HostNameOrIPAddress);
```

Comment:
This is a blocking call. Maybe it's permissible here? I think it would be better to make this method async and push the blocking all the way up the stack to (I think) `LocalSiloDetails`

Reply:
Yes I removed the async from the original method, since I didnd't see the point since it is used in `LocalSiloDetails`... And I don't see why we would use it elsewhere?